### PR TITLE
Fixed transport client leakage

### PR DESCRIPTION
**Pull request by jwillemsen**
DPFactoryQoS now doesn't report any leaked datareaders anymore, making a PR to get this to compile on CI

### CODE REVIEW

**Comment by jwillemsen** [comment]
With these changes and the changes from #642 combined I was able to run the DPFactoryQos test with RTPS discovery memory leak free.

**Comment by jwillemsen** [comment]
Reconnect test seems to fail because on_connection_deleted isn't called, but that is called from the destructor of TcpConnection which is also refcounted.

**Comment by jwillemsen** [comment]
Have been looking at the Reconnect test and I get the impression that we as DataWriter don't get the on_connection_deleted because we now cleanly detach from the transport before the event happens. Previously the transport kept the datawriter (with associated listener) alive. At the moment I remove the cleanup_transport from the DataWriter::cleanup() the test does run. Could it that we get the event now after the DataWriter cleanup has been called?

**Comment by mitza-oci** [comment]
retest this please
(starting Jenkins build)

**Comment by mitza-oci** [comment]
retest this please
(starting Jenkins build)

**Comment by mitza-oci** [comment]
retest this please
(starting other Jenkins build)

**Comment by jrw972** [comment]
retest this please
(starting other{2} Jenkins build)

**Comment by jrw972** [comment]
retest this please
(starting Jenkins build)

File: dds/DCPS/DiscoveryBase.h [hunk 1]
```diff
@@ -1224,7 +1221,10 @@ namespace OpenDDS {
         if (part == domain->second.end()) {
           return false;
         }
-        participant = part->second;
+        // Use reference counting to ensure participant
```
Comment (jwillemsen) [inline dds/DCPS/DiscoveryBase.h]: Removed the comment

**Comment by mitza-oci** [comment]
When working on PR #710 we found that an explicit method call such as TransportClient::cleanup_transport() was not necessary to fix leaks.  If there are still scenarios where this does help, please let us know details of which tests to run.

**Comment by jwillemsen** [comment]
Let us wait on the valgrind build, when #710 would fix the same issue we should see a decrease in test failures

**Comment by jwillemsen** [comment]
Did run some tests locally and the results are much better, good to see a different and more easier fix. Restarted the valgrind build to get fresh results

**Pull request closed without merging**